### PR TITLE
Update robotdrive for 2018

### PR DESCRIPTION
**Pull request by Twinters007**
* Deprecate RobotDrive
* Introduce DifferentialDrive, KilloughDrive, MecanumDrive, RobotDriveBase
* Update/validate motor controllers

Before dustin says something, I gotta do tests.

### CODE REVIEW

**Comment by Twinters007** [comment]
Robotdrive tests need a serious upgrade...I probably need help. @virtuald with the recent removal of the different constructors, and speedgroups instead of passing in the motors, the tests will need to update how it handles it all. Maybe you have some ideas since you made the original stuff?

**Comment by virtuald** [comment]
I can take a look this weekend probably.

File: wpilib/wpilib/robotdrive.py [hunk 1]
```diff
@@ -25,9 +25,12 @@ def _freeRobotDrive(allocatedSpeedControllers):
         if hasattr(sc, "free"):
             sc.free()
 
-
+@DeprecationWarning
```
Comment (auscompgeek) [inline wpilib/wpilib/robotdrive.py]: This isn't how warnings work.

File: wpilib/wpilib/robotdrive.py [hunk 2]
```diff
@@ -25,9 +25,12 @@ def _freeRobotDrive(allocatedSpeedControllers):
...
 class RobotDrive(MotorSafety):
-    """Operations on a robot drivetrain based on a definition of the motor
+    """
+    Deprecated. Use drive/* instead.
```
Comment (auscompgeek) [inline wpilib/wpilib/robotdrive.py]: I'm pretty sure there's a Sphinx `:deprecated:` thing or something.

**Comment by auscompgeek** [comment]
In terms of tests, we run the examples in https://github.com/robotpy/examples in our Travis tests as well. Would be worth porting across the new examples from allwpilib.

File: wpilib/wpilib/robotdrive.py [hunk 3]
```diff
@@ -25,10 +25,9 @@ def _freeRobotDrive(allocatedSpeedControllers):
         if hasattr(sc, "free"):
             sc.free()
 
-@DeprecationWarning
 class RobotDrive(MotorSafety):
     """
-    Deprecated. Use drive/* instead.
+    .. deprecated:: 2018.0.0. Use DifferentialDrive, KilloughDrive, or MecanumDrive instead.
```
Comment (auscompgeek) [inline wpilib/wpilib/robotdrive.py]: `. Use DifferentialDrive, KilloughDrive, or MecanumDrive instead.` isn't part of the version number.

File: wpilib/tests/test_robotdrive.py [hunk 4]
```diff
@@ -46,36 +34,32 @@ def test_init_two(kw, mc, wpimock, halmock):
...
+def test_init_four(kw, wpimock, halmock):
     halmock.getFPGATime.return_value = 1000
     halmock.getLoopTiming.return_value = wpimock.SensorBase.kSystemClockTicksPerMicrosecond
 
-    if mc is None:
-        fleft = MagicMock()
-        rleft = MagicMock()
-        fright = MagicMock()
-        rright = MagicMock()
     else:
```
Comment (auscompgeek) [inline wpilib/tests/test_robotdrive.py]: Nice syntax error you have there. (Did you actually try to run the tests before committing?)

File: wpilib/wpilib/robotdrive.py [hunk 5]
```diff
@@ -86,6 +88,7 @@ def __init__(self, *args, **kwargs):
         numbers are passed, the motorController keyword argument, if present,
         is the motor controller class to use; if unspecified, :class:`.Talon` is used.
         """
+        warnings.warn("Deprecated in 2018.0.0. Use DifferentialDrive, KilloughDrive, or MecanumDrive instead", DeprecationWarning)
```
Comment (auscompgeek) [inline wpilib/wpilib/robotdrive.py]: You should tell the user _what_ is deprecated, and where the deprecated code is being used (with the `stacklevel` argument).

File: wpilib/tests/test_robotdrive.py (hunk 4 above)
Comment (Twinters007) [reply wpilib/tests/test_robotdrive.py]: No I’m in the middle of it rn

File: wpilib/wpilib/drive/killoughdrive.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+                                math.sin(math.radians(leftMotorAngle)))
+        self.rightVec = Vector2d(math.cos(math.radians(rightMotorAngle)),
+                                 math.sin(math.radians(rightMotorAngle)))
+        self.backVec = Vector2d(math.cos(math.radians(backMotorAngle)),
+                                math.sin(math.radians(backMotorAngle)))
+
+        self.reported = False
+
+
+def driveCartesian(self, x, y, rotation, gyroAngle=0.0):
```
Comment (virtuald) [inline wpilib/wpilib/drive/killoughdrive.py]: Missing indentation.

File: wpilib/wpilib/__init__.py [hunk 7]
```diff
@@ -73,6 +74,8 @@
 
 from ._impl.main import run
 
+RobotDrive = DifferentialDrive
```
Comment (auscompgeek) [inline wpilib/wpilib/__init__.py]: ???

Comment (Twinters007) [reply wpilib/wpilib/__init__.py]: WIP

**Comment by Twinters007** [comment]
- [ ] Docstrings
- [ ] Tests

File: wpilib/wpilib/robotdrive.py [hunk 8]
```diff
@@ -86,6 +89,7 @@ def __init__(self, *args, **kwargs):
         numbers are passed, the motorController keyword argument, if present,
         is the motor controller class to use; if unspecified, :class:`.Talon` is used.
         """
+        warnings.warn("Deprecated in 2018.0.0. Use DifferentialDrive, KilloughDrive, or MecanumDrive instead.", DeprecationWarning, stacklevel=2)
```
Comment (auscompgeek) [inline wpilib/wpilib/robotdrive.py]: The user still can't really tell what exactly has been deprecated from this message.

Comment (Twinters007) [reply wpilib/wpilib/robotdrive.py]: Do you think it should be more like "RobotDrive was deprecated in 2018.0.0"?

**Comment by auscompgeek** [comment]
Something like "RobotDrive is deprecated, use XYZ instead" would suffice.

>

**Comment by virtuald** [comment]
This looks really good -- the tests seem quite comprehensive. Thanks!

**Comment by virtuald** [comment]
Also, if you're up to it... it would be useful to add pyfrc physics implementations for the new drives. :)

File: wpilib/wpilib/drive/__init__.py [hunk 9]
```diff
@@ -0,0 +1,5 @@
+from .differentialdrive import *
+from .killoughdrive import *
+from .mecanumdrive import *
+from .robotdrivebase import *
+from .vector2d import *
```
Comment (auscompgeek) [inline wpilib/wpilib/drive/__init__.py]: Just noticed this when I tested generating the docs (for no particular reason).

Do we really need/want Vector2d accessible as `wpilib.Vector2d`?

File: wpilib/wpilib/drive/vector2d.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+        """Rotate a vector in Cartesian space.
+
+        :param angle: Angle in degrees by which to rotate vector counter-clockwise
+        """
+        angle = math.radians(angle)
+        cosA = math.cos(angle)
+        sinA = math.sin(angle)
+
+        self.x = self.x * cosA - self.y * sinA
+        self.y = self.x * sinA + self.y * cosA
```
Comment (auscompgeek) [inline wpilib/wpilib/drive/vector2d.py]: This is wrong, very wrong, and doesn't match upstream. How did we not catch this?

Comment (virtuald) [reply wpilib/wpilib/drive/vector2d.py]: I would say it's subtly wrong from upstream, as only the last line is incorrect. Good catch.

File: wpilib/wpilib/drive/differentialdrive.py [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+
+        A differential drive robot has left and right wheels separated by an arbitrary width.
+
+        Drive base diagram:
+
+        |_______|
+        | |   | |
+          |   |
+        |_|___|_|
+        |       |
```
Comment (auscompgeek) [inline wpilib/wpilib/drive/differentialdrive.py]: Something gives me the feeling this was meant to be rendered preformatted. It isn't.

File: wpilib/wpilib/drive/__init__.py (hunk 9 above)
Comment (Twinters007) [reply wpilib/wpilib/drive/__init__.py]: Maybe? If teams want access to it outside of robotdrive classes?

File: wpilib/wpilib/drive/differentialdrive.py (hunk 11 above)
Comment (Twinters007) [reply wpilib/wpilib/drive/differentialdrive.py]: ?

File: wpilib/wpilib/drive/__init__.py (hunk 9 above)
Comment (auscompgeek) [reply wpilib/wpilib/drive/__init__.py]: I can't see a scenario where a team would want to use it over numpy…

**Pull request merged**